usecases/config/runtime/manager.go

```diff
@@ -0,0 +1,194 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package runtime
+
+import (
+	"context"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	ErrEmptyConfig         = errors.New("empty runtime config")
+	ErrFailedToOpenConfig  = errors.New("failed to open runtime config")
+	ErrFailedToReadConfig  = errors.New("failed to read runtime config ")
+	ErrFailedToParseConfig = errors.New("failed to parse runtime config ")
+)
+
+// Parser takes care of unmarshaling a config struct
+// from given raw bytes(e.g: YAML, JSON, etc).
+type Parser[T any] func([]byte) (*T, error)
```

Comment:
why does the parser return a pointer?

Reply:
nothing too specific. (1) Easy to assign it to config manager after parsing it. Instead of doing indirection. (2). In parser implementation, it's easy to return `nil` when parsing failed. Do you have any specific concern here?

Reply:
I just like to avoid pointers if possible because of lower risk of nil pointer access and less work for GC. But I guess here it does not matter that much